README
NAME

    HTTP::Tiny::PreserveHostHeader - preserve Host header on requests

SYNOPSIS

        use HTTP::Tiny::PreserveHostHeader;
    
        my $response = HTTP::Tiny::PreserveHostHeader->new->get(
            'http://example.com', {
                headers => {
                    Host => 'example.net',
                }
            }
        );

DESCRIPTION

    This module extends HTTP::Tiny and allows to preserve original Host
    header from HTTP request.

    The HTTP::Tiny is strictly compatible with HTTP 1.1 spec, section
    14.23:

      The Host field value MUST represent the naming authority of the
      origin server or gateway given by the original URL.

    It means that HTTP::Tiny always rewrite Host header to the value taken
    from URL.

    Some non-standard HTTP clients, such as reverse HTTP proxy, need to
    override Host header to other value.

SEE ALSO

    HTTP::Tiny, https://github.com/chansen/p5-http-tiny/pull/34.

BUGS

    If you find the bug or want to implement new features, please report it
    at https://github.com/dex4er/perl-HTTP-Tiny-PreserveHostHeader/issues

    The code repository is available at
    http://github.com/dex4er/perl-HTTP-Tiny-PreserveHostHeader

AUTHOR

    Piotr Roszatycki <dexter@cpan.org>

LICENSE

    Copyright (c) 2014-2016, 2023 Piotr Roszatycki <dexter@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as perl itself.

    See http://dev.perl.org/licenses/artistic.html



Changes
0.0103  2023-06-18      Piotr Roszatycki <dexter@cpan.org>

* Correct configure requirements.

0.0102  2023-06-17      Piotr Roszatycki <dexter@cpan.org>

* Use cpanfile for Module::Build.

0.0101  2023-06-11      Piotr Roszatycki <dexter@cpan.org>

* Provides cpanfile.

0.0100  2023-06-10      Piotr Roszatycki <dexter@cpan.org>

* First CPAN release.


Dist.SH
#!/usr/bin/env bash

set -e

PERL=${PERL:-perl}
PROVE=${PROVE:-prove}

test -x Build && ./Build distclean

${PERL} Build.PL

rm -f MANIFEST

${PERL} -ni -e 'print unless /^## NAME/ .. eof' README.md
dist_version_from=$(${PERL} -ne 'print if /dist_version_from/' Build.PL | ${PERL} -pe "s%.*=> '%%; s%',%%")
${PERL} -MPod::Markdown -e '$p=Pod::Markdown->new;$p->parse_from_filehandle(\*STDIN);print $p->as_markdown' \
    <"${dist_version_from:-lib/$(${PERL} -ne 'next unless /^name: /; s/^name: //; s,-,/,g; s/$/.pm/; print' META.yml)}" |
    ${PERL} -pe 'sub a { $_[0]=~tr/ /-/; $_[0] } s,[(]http://search.cpan.org/perldoc[?]([^#)]*)(#[^)]*)?[)],a("(http://metacpan.org/module/$1$2)"),eg' |
    ${PERL} -pe 's/^# /## /' \
        >>README.md

${PERL} Build manifest

${PERL} Build distdir
diff -u -I^SIGNATURE MANIFEST ./*-[0-9]*.[0-9]*/MANIFEST

cd ./*-[0-9]*.[0-9]*

${PERL} Build.PL
${PERL} Build

${PROVE} -Iblib/lib t

if [[ -z ${TEST_SKIP_EXTRA} ]] || [[ ${TEST_SKIP_EXTRA} == 0 ]]; then
    ${PROVE} -Iblib/lib xt
elif [[ -n ${TEST_SKIP_EXTRA} ]] && [[ ${TEST_SKIP_EXTRA} != 1 ]]; then
    # shellcheck disable=SC2046
    ${PROVE} -Iblib/lib $(find xt -name '*.t' ! -name "${TEST_SKIP_EXTRA}")
fi

if [[ $1 == sign ]]; then
    ${PERL} Build distsign
    diff -u -I^SIGNATURE MANIFEST ./*-[0-9]*.[0-9]*/MANIFEST
else
    ${PERL} Build distdir
fi

chmod -R u+w,go-w ./*-[0-9]*.[0-9]*

export COPYFILE_DISABLE=true
tar zcf "$(basename "$(pwd)").tar.gz" ./*-[0-9]*.[0-9]*

mv ./*.tar.gz ..


Build.PL
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use Module::Build;
use Module::CPANfile;

my $module = 'HTTP::Tiny::PreserveHostHeader';

(my $dist = $module) =~ s/::/-/g;

my $prereqs = Module::CPANfile->load->prereqs;

my $build = Module::Build->new(
    module_name             => $module,
    license                 => 'perl',
    build_requires          => $prereqs->requirements_for('build', 'requires')->as_string_hash,
    requires                => $prereqs->requirements_for('runtime', 'requires')->as_string_hash,
    recommends              => $prereqs->requirements_for('runtime', 'recommends')->as_string_hash,
    test_requires           => $prereqs->requirements_for('test', 'requires')->as_string_hash,
    auto_configure_requires => 0,
    create_makefile_pl      => 'traditional',
    create_readme           => 1,
    create_license          => 1,
    recursive_test_files    => 0,
    add_to_cleanup          => [
        '*.bak', '*.gz', '*.tgz', '*.ppd', '*.ppmx', '*.deb', '*.bat',
        'Makefile', 'pm_to_blib', 'cover_db', 'Debian_CPANTS.txt',
        "$dist-*",
    ],
    meta_merge => {
        resources => {
            bugtracker => "http://github.com/dex4er/perl-$dist/issues",
            repository => "http://github.com/dex4er/perl-$dist",
        },
    },
);

$build->create_build_script;


lib/HTTP/Tiny/PreserveHostHeader.pm
package HTTP::Tiny::PreserveHostHeader;

=head1 NAME

HTTP::Tiny::PreserveHostHeader - Preserve Host header on requests

=head1 SYNOPSIS

=for markdown ```perl

    use HTTP::Tiny::PreserveHostHeader;

    my $response = HTTP::Tiny::PreserveHostHeader->new->get(
        'http://example.com', {
            headers => {
                Host => 'example.net',
            }
        }
    );

=for markdown ```

=head1 DESCRIPTION

This module extends L<HTTP::Tiny> and allows to preserve original C<Host>
header from HTTP request.

The L<HTTP::Tiny> is strictly compatible with HTTP 1.1 spec, section 14.23:

=over

The Host field value MUST represent the naming authority of the origin
server or gateway given by the original URL.

=back

It means that L<HTTP::Tiny> always rewrite C<Host> header to the value
taken from URL.

Some non-standard HTTP clients, such as reverse HTTP proxy, need to override
C<Host> header to other value.

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0103';

use parent qw(HTTP::Tiny);

## no critic(Subroutines::ProhibitUnusedPrivateSubroutines)
sub _prepare_headers_and_cb {
    my ($self, $request, $args, $url, $auth) = @_;

    my $host;

    while (my ($k, $v) = each %{ $args->{headers} }) {
        if (lc $k eq 'host') {
            $host = $v;
            delete $args->{headers}{$k};
        }
    }

    $self->SUPER::_prepare_headers_and_cb($request, $args, $url, $auth);

    $request->{headers}{host} = $host if $host;

    return;
}

## no critic(Subroutines::ProhibitUnusedPrivateSubroutines)
sub _agent {
    my ($self) = @_;
    my $class = ref $self ? ref $self : $self;
    (my $default_agent = $class) =~ s{::}{-}g;
    ## no critic(Subroutines::ProtectPrivateSubs)
    return $default_agent . "/" . ($class->VERSION || 0) . " " . HTTP::Tiny->_agent;
}

1;

=for readme continue

=head1 SEE ALSO

L<HTTP::Tiny>, L<https://github.com/chansen/p5-http-tiny/pull/34>.

=head1 BUGS

If you find the bug or want to implement new features, please report it at
L<https://github.com/dex4er/perl-HTTP-Tiny-PreserveHostHeader/issues>

The code repository is available at
L<http://github.com/dex4er/perl-HTTP-Tiny-PreserveHostHeader>

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Copyright (c) 2014-2016, 2023 Piotr Roszatycki <dexter@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>
